### PR TITLE
Add `doctrine/orm` 3 support

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

This PR tries to add support for `doctrine/orm` 3

I am targeting this branch, because these changes are BC.

## Changelog

```markdown
### Added
- Support for `doctrine/orm` 3
```

The CI should be modified to allow installing `doctrine/orm` 3

### CODE REVIEW

File: src/AuditReader.php
Comment (franmomu) [inline src/AuditReader.php]: I had to add this because I was getting this kind of error:

```
Caused by
AssertionError: assert($this->association !== null) in EntityAuditBundle/vendor/doctrine/orm/lib/Doctrine/ORM/PersistentCollection.php:140
Stack trace:
#0 EntityAuditBundle/vendor/doctrine/orm/lib/Doctrine/ORM/PersistentCollection.php(140): assert(false, 'assert($this->a...')
#1 EntityAuditBundle/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php(508): Doctrine\ORM\PersistentCollection->hydrateAdd(Object(Sonata\EntityAuditBundle\Tests\Fixtures\Relation\OwnedEntity4))
#2 EntityAuditBundle/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/ObjectHydrator.php(143): Doctrine\ORM\Internal\Hydration\ObjectHydrator->hydrateRowData(Array, Array)
#3 EntityAuditBundle/vendor/doctrine/orm/lib/Doctrine/ORM/Internal/Hydration/AbstractHydrator.php(166): Doctrine\ORM\Internal\Hydration\ObjectHydrator->hydrateAllData()
#4 EntityAuditBundle/vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php(967): Doctrine\ORM\Internal\Hydration\AbstractHydrator->hydrateAll(Object(Doctrine\DBAL\Result), Object(Doctrine\ORM\Query\ResultSetMapping), Array)
#5 EntityAuditBundle/vendor/doctrine/orm/lib/Doctrine/ORM/Persisters/Entity/BasicEntityPersister.php(978): Doctrine\ORM\Persisters\Entity\BasicEntityPersister->loadCollectionFromStatement(Object(Doctrine\ORM\Mapping\ManyToManyInverseSideMapping), Object(Doctrine\DBAL\Result), Object(Doctrine\ORM\PersistentCollection))
#6 EntityAuditBundle/src/AuditReader.php(1184): Doctrine\ORM\Persisters\Entity\BasicEntityPersister->loadManyToManyCollection(Object(Doctrine\ORM\Mapping\ManyToManyInverseSideMapping), Object(Sonata\EntityAuditBundle\Tests\Fixtures\Relation\OwnerEntity), Object(Doctrine\ORM\PersistentCollection))
#7 EntityAuditBundle/src/AuditReader.php(328): SimpleThings\EntityAudit\AuditReader->createEntity('Sonata\\EntityAu...', Array, Array, 1)
#8 EntityAuditBundle/tests/RelationTest.php(349): SimpleThings\EntityAudit\AuditReader->find('Sonata\\EntityAu...', Array, 1)
#9 EntityAuditBundle/vendor/phpunit/phpunit/src/Framework/TestCase.php(1612): Sonata\EntityAuditBundle\Tests\RelationTest->testManyToMany()
```
cc @greg0ire just in case you want to take a look

**Comment by franmomu** [comment]
We need https://github.com/doctrine/data-fixtures/pull/452

File: src/AuditReader.php
Comment (greg0ire) [reply src/AuditReader.php]: @franmomu thanks for letting me know, I opened https://github.com/doctrine/orm/pull/11071 to get that reverted.

**Comment by greg0ire** [comment]
The next blocker seems to be https://github.com/doctrine-extensions/DoctrineExtensions/issues/2708

**Comment by franmomu** [comment]
> The next blocker seems to be [doctrine-extensions/DoctrineExtensions#2708](https://github.com/doctrine-extensions/DoctrineExtensions/issues/2708)

Yep, that would take a bit longer I think, I'll try to help with that

**Comment by VincentLanglet** [comment]
Hi @franmomu, what about following the same strategy than https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/4.x/.github/workflows/test_orm_3.yaml ?

gedmo/doctrine-extensions is an optional dev dependency, it should be a blocker for the ORM 3 support.

**Review by ghost — APPROVED** [review]

**Comment by ghost** [comment]
Is there anything I can do to help this PR along? This bundle is holding up an ORM upgrade on our end, so I'm happy to help if I can.

**Comment by VincentLanglet** [comment]
> Is there anything I can do to help this PR along? This bundle is holding up an ORM upgrade on our end, so I'm happy to help if I can.

Basically, we need green tests.

If you want you can recreate the same PR and add a github actions similar to https://github.com/sonata-project/SonataDoctrineORMAdminBundle/blob/4.x/.github/workflows/test_orm_3.yaml

Instead of removing AuditBundle you remove the doctrineExtensions.

**Comment by franmomu** [comment]
I found some time to try to work on this, I've tried to remove some deprecations (related to accessing new mapping objects as array), but there should be more.

Not sure if I can keep working on this in the following days, so if anyone wants to continue with this, feel free to do it.

**Review by ghost — APPROVED** [review]

**Comment by franmomu** [comment]
I think this is ready to be reviewed, the `SimpleThings\EntityAudit\Utils\ORMCompatibilityTrait` tries to avoid deprecations triggered by ORM 3 accessing to mapping objects as array.

At first, I've used a `getMappingValue` method for almost everything, and then I've starting using specific methods (`getMappingColumnNameValue`, `getMappingNameValue`, etc) to be able to be more specific about types.

File: src/AuditReader.php
Comment (VincentLanglet) [inline src/AuditReader.php]:
```
Doctrine\ORM\Exception\ORMException
```
already exist in doctrine/orm 2.

And since `ORMException` extends `ORM2Exception` exception, I assume the doctrine codebase was updated and already throw `ORMException` everywhere ; so I would say we can remove all the `@throws ORM2Exception` tag, wdyt @franmomu ?

File: src/AuditReader.php
Comment (franmomu) [inline src/AuditReader.php:260]: The `joinColumnFieldNames` [is only available](https://github.com/search?q=repo%3Adoctrine%2Form%20joinColumnFieldNames&type=code) in [`ToOneOwningSideMapping`](https://github.com/doctrine/orm/blob/1153b9468ccfef452625064c66b4a135d0ea8a46/src/Mapping/ToOneOwningSideMapping.php#L26)

File: src/AuditReader.php
Comment (franmomu) [inline src/AuditReader.php:427]: The `targetToSourceKeyColumns` field [is only available](https://github.com/search?q=repo%3Adoctrine%2Form+targetToSourceKeyColumns&type=code) in [`ToOneOwningSideMapping`](https://github.com/doctrine/orm/blob/1153b9468ccfef452625064c66b4a135d0ea8a46/src/Mapping/ToOneOwningSideMapping.php#L20)

File: src/AuditReader.php
Comment (franmomu) [reply src/AuditReader.php]: Ah didn't check that, changed, thanks

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Do you want me to merge/release or should we wait ? @franmomu

**Comment by franmomu** [comment]
since I've change some things, it would be nice to have another approval, I've requested reviews 🤞

**Review by greg0ire — APPROVED** [review]

File: src/AuditConfiguration.php
Comment (phansys) [inline src/AuditConfiguration.php:82]: ~~As this class is not final or marked as internal, I think this is technically a BC break.
Maybe we can live with this documenting this decision.~~

Comment (phansys) [reply src/AuditConfiguration.php:82]: My bad (I was confusing `Doctrine\ORM\Mapping\ClassMetadata` with `Doctrine\Persistence\Mapping\ClassMetadata`).

**Review by phansys — APPROVED** [review]

**Comment by phansys** [comment]
Thank you @franmomu!

**Comment by VincentLanglet** [comment]
Go for the release https://github.com/sonata-project/EntityAuditBundle/pull/628

**Pull request merged**
